### PR TITLE
Simplify ocall stack stitching

### PR DESCRIPTION
**Pull request by anakrish**
PR to get early feedback.

OE SDK performs stack stitching for ocalls (debug enclaves only).
This drastically simplifies the debugger implementations
-  frees the debuggers from knowing about any OE specific data
 structures on the enclave side.
- debuggers no longer have to perform stack stitching

### CODE REVIEW

File: enclave/core/sgx/enter.S [hunk 1]
```diff
@@ -188,10 +194,10 @@ oe_enter:
     push %rdi
     push %rsi
 
-    // Notify a nested exit happens.
-    // oe_notify_nested_exit_start(rdi=arg1, rsi=ocall_context)
+    // For debug enclave, save the eexit frame in host memory.
+    // oe_save_eexit_frame_in_host_memory(rdi=arg1, rsi=eexit_frame_info)
     mov %r12, %rsi
-    call oe_notify_nested_exit_start
+    call oe_save_eexit_frame_in_host_memory
```
Comment (radhikaj) [inline enclave/core/sgx/enter.S]: How are you deciding if its a debug enclave or not before calling this function?

File: enclave/core/sgx/calls.c [hunk 2]
```diff
@@ -856,40 +857,61 @@ void __oe_handle_main(
...
-**     exit).
-**     When debugger does stack stitching, it will update the untrusted ocall
-**     frame's previous stack frame pointer and return address with the ocall
-**     context from trusted thread data. When GDB does stack walking, the parent
-**     stack of an untrusted ocall will be stack of the _OE_EXIT trusted
-**     function instead of stack of oe_enter/__morestack untrusted function.
-**     Refer to the oe_notify_ocall_start function in host side, and the
-**     OCallStartBreakpoint and update_untrusted_ocall_frame function in the
-**     python plugin.
+**     This function saves the current eexit frame in host specified location.
```
Comment (dthaler) [inline enclave/core/sgx/calls.c]: grammar: "in host specified" -> "in a host-specified"

File: enclave/core/sgx/calls.c [hunk 3]
```diff
@@ -856,40 +857,61 @@ void __oe_handle_main(
...
+**     This allows the host to stitch the stack, instead of the debugger as
+**     was done before. The frame is save only for debug enclaves.
```
Comment (dthaler) [inline enclave/core/sgx/calls.c]: "save" -> "saved"?

File: include/openenclave/internal/sgxtypes.h [hunk 4]
```diff
@@ -565,10 +565,16 @@ oe_thread_data_t* oe_get_thread_data(void);
 
 #define TD_MAGIC 0xc90afe906c5d19a3
 
-#define OE_THREAD_LOCAL_SPACE (3840)
+#define OE_THREAD_LOCAL_SPACE (3832)
```
Comment (dthaler) [inline include/openenclave/internal/sgxtypes.h]: Please add a comment saying how the number 3832 was calculated

File: tests/debugger/oegdb/enc/contract.c [hunk 5]
```diff
@@ -8,36 +8,12 @@
...
-// volatile is used to prevent compiler-optimizations.
-volatile uint64_t TD_OFFSET_FROM_TCS = (uint64_t)-1;
-volatile uint64_t TD_CALLSITE_OFFSET = (uint64_t)-1;
-volatile uint64_t CALLSITE_OCALLCONTEXT_OFFSET = (uint64_t)-1;
-volatile uint64_t OCALLCONTEXT_LENGTH = (uint64_t)-1;
-volatile char OCALLCONTEXT_FORMAT[10];
-volatile uint64_t OCALLCONTEXT_RBP = (uint64_t)-1;
-volatile uint64_t OCALLCONTEXT_RET = (uint64_t)-1;
-
+// The debugger no loger relies on any OE specific structures on enlave side.
```
Comment (dthaler) [inline tests/debugger/oegdb/enc/contract.c]: typo "enlave".   (Also suggest spelling out OE.)

File: tests/debugger/oegdb/enc/contract.c [hunk 6]
```diff
@@ -8,36 +8,12 @@
...
+// The stack stitching for ocalls is done by the SDK itself.
+// Note: The debugger uses SGX defined sgx_tcs_t for setting the debug-optin
```
Comment (dthaler) [inline tests/debugger/oegdb/enc/contract.c]: to confirm, did you mean "optin" or "option"?

**Comment by anakrish** [comment]
This will be done at a later time.

**Pull request closed without merging**